feat: splash screen support for Electron

sdirix opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md

Note: Security vulnerabilities should not be disclosed on GitHub, through a PR or any
other means. See SECURITY.md at the root of this repository, to learn how to report
vulnerabilities.
-->

#### What it does

Enhances the `ElectronMainApplication` to optionally render a splash screen until the frontend is ready.

The splash screen can be configured via the application config object `theia.frontend.config.electron.splashScreenOptions`. Mandatory is the option `content` which specifies a relative path from the frontend location to the content of the splash screen. Optionally `width`, `height`, `minDuration` and `maxDuration` can be handed over too.

By default the splash screen is off.

Implements #13410

Contributed on behalf of Pragmatiqu IT GmbH

[TheiaSplashScreen.webm](https://github.com/eclipse-theia/theia/assets/8998368/e0c95070-d1dc-46ee-8563-81545f1aa5af)


#### How to test

1. Place a png in `examples/electron/resources/` directory
2. Modify the Electron example application `package.json` to configure the splash screen, e.g. like this
    ```
    "theia": {
      "target": "electron",
      "frontend": {
        "config": {
          "applicationName": "Theia Electron Example",
          "reloadOnReconnect": true,
          "electron": {
            "splashScreenOptions": {
              "content": "resources/my-logo.png"
            }
          }
        }
      },
      "backend": {
        "config": {
          "frontendConnectionTimeout": -1
        }
      }
    }
   ```

3. Build and start the Electron example app

Play around with the various options, e.g. hand over a `html` file as the content and try the configuration options `width`, `height`, `minDuration` and `maxDuration`.

The PR now also configured the Electron example app already

#### Review checklist

- [X] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

tsmaeder commented on packages/core/src/electron-main/electron-main-application.ts:
```diff
@@ -287,12 +296,69 @@ export class ElectronMainApplication {
+                    frame: false,
+                    alwaysOnTop: true,
+                    webPreferences: {
+                        preload: content.endsWith('.js') ? path.resolve(this.globals.THEIA_APP_PROJECT_PATH, 'lib', 'frontend', content).toString() : undefined
```
Wouldn't `require.resolve()` be a more robust way to load a javascript file?

tsmaeder replied:
Also, why are we loading `*.js` files in preload and not like the other files?

sdirix replied:
We can't load Javascript files like regular documents as then we will show their content instead of executing them. We would need to specifically call `executeJavascript` and load it beforehand.

Instead I wanted to support the use case of executing within the preload to allow for customizations like transferring information for a progress bar. However as we don't load anything else it's then difficult to get actual content into the window without customizing `showSplashScreen` itself anyway.

We could offer both `content` and `preload` customizations, but that's maybe overkill. `content` should be enough and I'll remove the preload handling. If an adopter needs more control over the window they can still customize `showSplashScreen`.

tsmaeder commented on dev-packages/application-package/src/application-props.ts:
```diff
@@ -33,8 +33,37 @@ export type ElectronFrontendApplicationConfig = RequiredRecursive<ElectronFronte
+         *
+         * Mandatory attribute.
+         */
+        content?: string;
```
Also, if it's documented mandatory, but typed as optional.

sdirix replied:
The type is optional as we simply don't know what the user puts into the actual configuration. However if it's not there we turn off the splash screen feature.

sdirix commented on dev-packages/application-package/src/application-props.ts:
```diff
@@ -45,11 +74,16 @@ export namespace ElectronFrontendApplicationConfig {
+         * Whether or not to show an empty Electron main window as early as possible.
          *
-         * Defaults to `true`.
+         * Has no effect if `splashScreenOptions.content` is also given.
```
good idea, that will clarify the functionality automatically from an adopter point of view.

Initially I thought we could support both, `showWindowEarly` and `splashScreenOptions` by showing the main window and the splash screen in case that `showWindowEarly` is true. However as it's true by default, almost all adopters which want to use the splash screen then would need to turn it off, which is why I did not go forward with it.

tsmaeder commented:
> Would you like to see it configured in the example?

Yes, I think that would neatly document how it's done.

sdirix commented:
> > Would you like to see it configured in the example?
> 
> Yes, I think that would neatly document how it's done.

Done :+1:

tsmaeder commented on packages/core/src/electron-main/electron-main-application.ts:
```diff
@@ -287,18 +300,87 @@ export class ElectronMainApplication {
+
+    protected determineSplashScreenBounds(): { x: number, y: number, width: number, height: number } {
+        const splashScreenOptions = this.getSplashScreenOptions();
+        const width = splashScreenOptions?.width ?? 640;
```
640x480? Is VGA still a thing? But seriously: I was wondering why we always have to give a size for the window. Does the browser window not adjust automatically to the size of the loaded file?

sdirix replied:
As we don't know what's in the content we can't really center it on screen without using some predetermined width/height. I'm open for other defaults or if you know some better way of choosing on which screen the splash shall appear without using x, y, width and height.

tsmaeder replied:
Does the window not auto-resize to fit the content? Maybe we can load the file and then query the window dimensions and position the window accordingly and only then do the `show`? https://www.electronjs.org/docs/latest/api/browser-window#showing-the-window-gracefully. 640x480 seems so arbitrary.

sdirix replied:
Generally speaking in the web it works the other way around, the content fits to the browser size.

In theory we could add some sort of analyzing to try to determine the content width. I guess we could use `window.webContents.executeJavascript(javascriptAsString)` and then try to implement some heuristic size guessing, falling back to our default size in the cases where we can't. I guess the only sizes where it's 100% sure are the ones where the content is really just a png or someone who set a max width (and max height) in their styles.

The downside of this approach is that it obstructs what the splash screen is for: To be shown as early as possible to indicate to the user that the app is loading. Instead of showing it as early as possible, we first need to render it non-visible, then execute the analysis Javascript, move the window into the correct position, resize and only then show it.

Alternative suggestion: Instead of hard coding the splash screen size default we could calculate the default based on the screen size, similar to the Theia window size default. Would you prefer that solution?

tsmaeder reviewed (commented):
One last comment.

sdirix commented:
>What don't we just make the type of showWindowEarly to be boolean | SplashScreenOptions? This would make the state space of the application simpler.

I thought about this: Maybe it would actually be better to split them. And `showWindowEarly` then controls whether we show the splash screen / main window before or after the window is ready to show content. I think that makes it more explicit and is actually a sensible option for both use cases.

tsmaeder commented:
> showWindowEarly then controls whether we show the splash screen / main window before or after the window is ready to show content.

Why configure a splash screen if you don't show it? So `splashScreeen/false` is a combination that never makes sense.

sdirix commented:
> > showWindowEarly then controls whether we show the splash screen / main window before or after the window is ready to show content.
> 
> Why configure a splash screen if you don't show it? So `splashScreeen/false` is a combination that never makes sense.

What I meant is:
 - `showWindowEarly` currently controls whether we show the main window early (i.e. immediately) or after the window sends the `ready-to-show` message
 - So we can keep this behavior also for the splash screen, i.e. in case we show a splash screen, the `showWindowEarly` should be interpreted for the splash screen, not for the main window. So then depending on the `showWindowEarly` we show the splash screen before or after the `ready-to-show` message.

I think that makes sense as some users might dislike the blank window which is we why allow to configure `showWindowEarly` in the first case.

tsmaeder commented:
@sdirix I would be fine to merge this one with the behavior as it is in order to have it in the release. However, I would rather not have adopters change the configuration twice in short succession, so if you want to change the config syntax, I would rather have it in this PR. The pre-release window is closing 🤷 .

sdirix commented:
I added the discussed changes and split the options again. I slightly adjusted the splash screen location determination to consider the center point of the Theia application. Before when Theia was closed with only 5 pixels on the left of two screens, the splash screen would be shown left instead of the right.

sdirix commented:
I noticed a regression during testing and refactored the code. The issue was that I previously added the `preventAutomaticShow` to the default window options. That's fine for the first window, but not for all other windows which are created later, e.g. via the "New Window" action.

I now only add the `preventAutomaticShow` when we actually are currently showing (or intending to show) a splash screen. This prevents the issue.
To avoid breaking changes I also reverted the change to `getLastWindowOptions` which I had refactored from async to sync. However if someone actually relies on that being async for their customizations, their code would break.

I think the code is now cleaner than before. It works well for me in testing.

tsmaeder commented:
@sdirix btw.: usually, we leave updates to the PR in separate commits until merge time so that reviewers don't have to review everything again when comments are being addressed.

tsmaeder commented:
Sorry about this being a bit late, but doing another pass got me thinking:

1. Taking `showWindowEarly` and `splashScreenOptions` apart has made things more complicated, IMO. 
2. Does it even make sense to show an empty Window if we have splash screen support?
3. If we always created the initial Window early, we could immediately set up the splash-screen handling in `showSplashScreen`, instead of potentially adding it to every new window in `configureShowOnSplashScreenClose()`

sdirix commented on packages/core/src/electron-main/electron-main-application.ts:
```diff
@@ -307,6 +405,10 @@ export class ElectronMainApplication {
     async createWindow(asyncOptions: MaybePromise<TheiaBrowserWindowOptions> = this.getDefaultTheiaWindowOptions()): Promise<BrowserWindow> {
         let options = await asyncOptions;
         options = this.avoidOverlap(options);
+        if (this.splashScreenState) {
```
`isShowSplashScreen()` will return `true` for the lifetime of the application as it's just about the configuration of a splash screen. `this.splashScreenState` only exist during the time we actually want to show a splash screen

sdirix commented on packages/core/src/electron-main/electron-main-application.ts:
```diff
@@ -287,9 +301,82 @@ export class ElectronMainApplication {
+            this.splashScreenState = {};
+            app.whenReady().then(() => {
+                this.determineSplashScreenBounds().then(splashScreenBounds => {
+                    const splashScreenWindow = new BrowserWindow({
```
Don't have hard feelings here, can we do as you want.

sdirix commented on packages/core/src/electron-main/electron-main-application.ts:
```diff
@@ -287,9 +301,82 @@ export class ElectronMainApplication {
+
+    protected isShowWindowEarly(): boolean {
+        return !!this.config.electron.showWindowEarly &&
+            !('THEIA_ELECTRON_NO_EARLY_WINDOW' in process.env && process.env.THEIA_ELECTRON_NO_EARLY_WINDOW === '1');
```
This is pre-existing code which I just moved into this function as I needed to call it again. I can refactor it if you want

sdirix commented on packages/core/src/electron-main/theia-electron-window.ts:
```diff
@@ -37,6 +37,12 @@ export interface TheiaBrowserWindowOptions extends BrowserWindowConstructorOptio
+     * This can be prevented by handing over preventAutomaticShow: `true`.
+     * Use this for fine-grained control over when to show the window, e.g. to coordinate with a splash screen.
+     */
+    preventAutomaticShow?: boolean;
```
>Do we really need to prevent the main window from showing as soon as it's ready? I don't see the use case and it makes the window state graph more complicated.

The splash screen was intended as an alternative to the loading spinner of the main window. If we don't prevent the automatic show of the main window we will see the splash screen as well as the full main window at the same time.

sdirix replied:
> Why not add an "open" call to TheiaElectronWindow and inoke it from the ElectronMainApplication class?

Sadly, I don't understand what you mean. What is the purpose of that open and when shall it be called?

tsmaeder replied:
The idea would be that the ElectronMainApplication controls visibility of the windows instead of passing flags to the window and letting the window manage it's own visiblity.

sdirix commented on packages/core/src/electron-main/electron-main-application.ts:
```diff
@@ -316,9 +418,40 @@ export class ElectronMainApplication {
+            });
+            // maxTime should always exist here, sanity fallback
+            const maxTime = this.splashScreenState.maxTime ?? new Promise(resolve => setTimeout(() => resolve(), 30000));
+            maxTime.then(() => showWindowAndCloseSplashScreen());
```
In my opinion we should always show *something*. So if the splash screen is force closed via maxTime, we show the current state of the main window.

sdirix commented:
> Sorry about this being a bit late, but doing another pass got me thinking:

No worries. We should get this right.
 
> 1. Taking `showWindowEarly` and `splashScreenOptions` apart has made things more complicated, IMO.

The code is more logical in my opinion. These are now just two separate options. `showWindowEarly` is about whether the first rendered window shall be shown as early as possible, even when this means looking at a blank window. 

`splashScreenOptions` is just for configuring optional splash screen content.

> 2. Does it even make sense to show an empty Window if we have splash screen support?

Yes. There are proponents for and against splash screens. We can't force everyone to use a splash screen for earliest possible visual feedback

> 3. If we always created the initial Window early, we could immediately set up the splash-screen handling in `showSplashScreen`, instead of potentially adding it to every new window in `configureShowOnSplashScreenClose()`

That's a good point. We could create the window early but just not show it. 

The reason I added it to `createWindow` thinking about use cases where multiple windows might be opened while the splash screen is shown, however that might be too much of an edge case to be worth it.

tsmaeder commented on packages/core/src/electron-main/electron-main-application.ts:
```diff
@@ -287,18 +288,110 @@ export class ElectronMainApplication {
+            }
+            splashScreenWindow.close();
+        };
+        TheiaRendererAPI.onApplicationStateChanged(mainWindow.webContents, state => {
```
Is this the right event or should we just wait until the browser window has finished loading it's content?

sdirix replied:
The idea is to wait until the frontend is ready to show something. Otherwise the splash screen will just be shown less than 500ms before being replaced by the main window showing a loading spinner.

sdirix commented:
>The idea would be that the ElectronMainApplication controls visibility of the windows instead of passing flags to the window and letting the window manage it's own visiblity.

Fine with me, I can refactor this

sdirix commented:
> > The idea would be that the ElectronMainApplication controls visibility of the windows instead of passing flags to the window and letting the window manage it's own visiblity.
> 
> Fine with me, I can refactor this

I had a look at this and I'm not sure we're gaining much. The issue is that we can't handle the "automatic" opening within the `createWindow` function by itself as we're lacking context there whether we want to open the window or not. So we would either need to a new parameter to `createWindow` to indicate whether the window shall be automatically opened (essentially exchanging one new parameter with another) or adapt all `createWindow` callers. There are three callers, so that would be manageable, but all three need to be adapted then with this additional logic. Additionally `createWindow` is public API, so if any consumer is using that already for their use cases, we will break them.

If you still prefer that we adapt the logic of the `BrowserWindow` then let me know.

sdirix commented:
Note that the Playwright tests are currently failing in all PRs, so this is unrelated to this PR.
